<a href="https://colab.research.google.com/github/redinbluesky/fun-coding-python-data-analysis/blob/main/07-실습:영화_데이터_분석하기.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  목차
* [Chapter 7-1 데이터 불러오기](#chapter7-1)
* [Chapter 7-2 데이터 확인하기](#chapter7-2)
    * [Chapter 7-2-1 데이터 구조파악](#chapter7-2-1)
    * [Chapter 7-2-2 기본 통계 정보 확인히가](#chapter7-2-2)
    * [Chapter 7-2-3 데이터 크기 확인하기](#chapter7-2-3)
* [Chapter 7-3 영화 평점과 참여주 수 분석하기](#chapter7-3)    
    * [Chapter 7-3-1 영화 평점의 평균 구하기](#chapter7-3-1)    
    * [Chapter 7-3-2 최고 평점 영화 찾기](#chapter7-3-2)    
    * [Chapter 7-3-3 최저 평점 영화 찾기](#chapter7-3-3)    
    * [Chapter 7-3-4 전체 평가 참여자 수와 영화당 평균 참여자 수 구하기](#chapter7-3-4)    
    * [Chapter 7-3-5 데이터 분석 결과](#chapter7-3-5)    
 * [Chapter 7-4 셀프체크](#chapter7-4)     

## Chapter 7-1 데이터 불러오기 <a class="anchor" id="chapter7-1"></a>
1. 데이터 분석의 첫 단계는 데이터셋을 불러와 각 열의 의미와 값을 정확히 파악하는 것이다. 
    - 이 과정을 탐색적 데이터 분석이라고 한다.(EDA: Exploratory Data Analysis)
    - 데이터 구조 이해: 각 열이 어떤 데이터를 담고 있는지 파악한다.
    - 데이터 품질 확인: 결측치, 이상치, 중복값 여부를 확인한다.
    - 페턴 및 관계 탐색: 변수 간 관계를 파악하고, 데이터에서 숨겨진 패턴을 찾는다.
    - EDA를 소홀히 하는 경우 데이터의 특성을 제대로 이해하지 못해 분석이 왜곡되거나 잘못된 결론에 도달할 수 있다.

2. IMDb(Internet Movie Database)에 있는 상위, 1,000개 영화와 TV 쇼에 대한 데이터를 이 장에서 사용한다.
    - imdb_top_1000.csv

3. 캐글(Kaggle)은 데이터 과학과 머신러닝 분야에서 가장 유명한 온라인 플랫폼이다.
    - 자신이 소유한 데이터를 업로드해 다른 사람과 공유할 수 있다.

In [2]:
# read_csv() 함수를 하용해 imdb_top_1000.csv 파일을 읽어온다.
import pandas as pd

# CSV 파일 경로
file_path = "./data/imdb_top_1000.csv"
# CSV 파일 불러오기
df = pd.read_csv(file_path)

## Chapter 7-2 데이터 확인하기 <a class="anchor" id="chapter7-2"></a>
### Chapter 7-2-1 데이터 구조파악 <a class="anchor" id="chapter7-2-1"></a>
1. IMBD 각 행은 영화 한편에 대한 정보를 담고 있다.
    - Poster_Link: 영화 포스터 이미지 링크
    - Series_Title: 영화 제목
    - Released_Year: 영화 개봉 연도
    - Certificate: 영화 등급
    - Runtime: 영화 상영 시간(분)
    - Genre: 영화 장르
    - IMDB_Rating: 영화 평점
    - Overview: 영화 줄거리
    - Meta_score: 영화 메타스코어
    - Director: 영화 감독
    - Star1: 영화 주연 배우 1
    - Star2: 영화 주연 배우 2
    - Star3: 영화 주연 배우 3
    - Star4: 영화 주연 배우 4
    - No_of_Votes: 영화에 대한 투표 수
    - Gross: 영화 수익(달러)

In [2]:
# head() 메서드를 사용하여 처음 5행 확인
df.head()

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
0,https://m.media-amazon.com/images/M/MV5BMDFkYT...,The Shawshank Redemption,1994,A,142 min,Drama,9.3,Two imprisoned men bond over a number of years...,80.0,Frank Darabont,Tim Robbins,Morgan Freeman,Bob Gunton,William Sadler,2343110,"28,341,469"
1,https://m.media-amazon.com/images/M/MV5BM2MyNj...,The Godfather,1972,A,175 min,"Crime, Drama",9.2,An organized crime dynasty's aging patriarch t...,100.0,Francis Ford Coppola,Marlon Brando,Al Pacino,James Caan,Diane Keaton,1620367,"134,966,411"
2,https://m.media-amazon.com/images/M/MV5BMTMxNT...,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,When the menace known as the Joker wreaks havo...,84.0,Christopher Nolan,Christian Bale,Heath Ledger,Aaron Eckhart,Michael Caine,2303232,"534,858,444"
3,https://m.media-amazon.com/images/M/MV5BMWMwMG...,The Godfather: Part II,1974,A,202 min,"Crime, Drama",9.0,The early life and career of Vito Corleone in ...,90.0,Francis Ford Coppola,Al Pacino,Robert De Niro,Robert Duvall,Diane Keaton,1129952,"57,300,000"
4,https://m.media-amazon.com/images/M/MV5BMWU4N2...,12 Angry Men,1957,U,96 min,"Crime, Drama",9.0,A jury holdout attempts to prevent a miscarria...,96.0,Sidney Lumet,Henry Fonda,Lee J. Cobb,Martin Balsam,John Fiedler,689845,"4,360,000"


In [3]:
# tail() 메서드를 사용하여 마지막 5행 확인
df.tail()

,Poster_Link,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Overview,Meta_score,Director,Star1,Star2,Star3,Star4,No_of_Votes,Gross
995,https://m.media-amazon.com/images/M/MV5BNGEwMT...,Breakfast at Tiffany's,1961,A,115 min,"Comedy, Drama, Romance",7.6,A young New York socialite becomes interested ...,76.0,Blake Edwards,Audrey Hepburn,George Peppard,Patricia Neal,Buddy Ebsen,166544,NaN
996,https://m.media-amazon.com/images/M/MV5BODk3Yj...,Giant,1956,G,201 min,"Drama, Western",7.6,Sprawling epic covering the life of a Texas ca...,84.0,George Stevens,Elizabeth Taylor,Rock Hudson,James Dean,Carroll Baker,34075,NaN
997,https://m.media-amazon.com/images/M/MV5BM2U3Yz...,From Here to Eternity,1953,Passed,118 min,"Drama, Romance, War",7.6,"In Hawaii in 1941, a private is cruelly punish...",85.0,Fred Zinnemann,Burt Lancaster,Montgomery Clift,Deborah Kerr,Donna Reed,43374,"30,500,000"
998,https://m.media-amazon.com/images/M/MV5BZTBmMj...,Lifeboat,1944,NaN,97 min,"Drama, War",7.6,Several survivors of a torpedoed merchant ship...,78.0,Alfred Hitchcock,Tallulah Bankhead,John Hodiak,Walter Slezak,William Bendix,26471,NaN
999,https://m.media-amazon.com/images/M/MV5BMTY5OD...,The 39 Steps,1935,NaN,86 min,"Crime, Mystery, Thriller",7.6,A man in London tries to help a counter-espion...,93.0,Alfred Hitchcock,Robert Donat,Madeleine Carroll,Lucie Mannheim,Godfrey Tearle,51853,NaN


### Chapter 7-2-2 기본 통계 정보 확인하기 <a class="anchor" id="chapter7-2-2"></a>
1. 데이터셋의 기본 통계 정보를 확인하는 것은 데이터 분석의 중요한 단계이다.
    - 데이터셋의 크기, 열의 이름과 데이터 타입, 결측치 여부 등을 파악할 수 있다.
    - 이를 통해 데이터의 품질을 평가하고, 이후 분석에서 어떤 전처리 작업이 필요한지 결정할 수 있다.

2. describe() 메서드를 사용해 기본 정보를 확인할 수 있다.
    - count: 각 열의 유효한 값의 개수
    - mean: 각 열의 평균값
    - std: 각 열의 표준편차
    - min: 각 열의 최솟값
    - 사분위 점수: 25%, 50%, 75%에 해당하는 값
    - max: 각 열의 최댓값 

3. 표준편차 해석방법
    - 표준편차는 데이터가 평균에서 얼마나 퍼져 있는지를 나타내는 지표이다.
    - 표준편차가 크면 데이터가 평균에서 멀리 퍼져 있고, 표준편차가 작으면 데이터가 평균에 가까이 모여 있음을 의미한다
        - 예를 들어, IMDB_Rating 열의 표준편차가 0.5라면, 대부분의 영화 평점이 평균에서 0.5점 이내에 분포하고 있음을 나타낸다.
    - 데이터의 분포 범위에 대한 이해를 돕는다.
        - 평균 +-1 표준편차 범위 내에 데이터의 약 68%가 포함되고, 평균 +-2 표준편차 범위 내에 약 95%가 포함된다는 점을 활용할 수 있다.

In [4]:
# descirbe() 메서드를 사용해 IMBD의 기본 통계 정보 확인
df.describe()

,IMDB_Rating,Meta_score,No_of_Votes
count,1000.000000,843.000000,1.000000e+03
mean,7.949300,77.971530,2.736929e+05
std,0.275491,12.376099,3.273727e+05
min,7.600000,28.000000,2.508800e+04
25%,7.700000,70.000000,5.552625e+04
50%,7.900000,79.000000,1.385485e+05
75%,8.100000,87.000000,3.741612e+05
max,9.300000,100.000000,2.343110e+06


### Chapter 7-2-3 데이터 크기 확인하기 <a class="anchor" id="chapter7-2-3"></a>

In [ ]:
# shape 속성을 사용해 데이터 크기 확인하기
#   - 1,000개의 행과 16개의 열을 가지고 있다.
df.shape

(1000, 16)

## Chapter 7-3 영화 평점과 참여주 수 분석하기 <a class="anchor" id="chapter7-3"></a>
### Chapter 7-3-1 영화 평점의 평균 구하기 <a class="anchor" id="chapter7-3-1"></a>


In [8]:
# IMDB 영화 평점의 평균 계산
average_rating = df['IMDB_Rating'].mean()
print("평점 평균: ", average_rating)

평점 평균:  7.949299999999999


### Chapter 7-3-2 최고 평점 영화 찾기 <a class="anchor" id="chapter7-3-2"></a>

In [ ]:
# IMDB 영화 중 최고 평점 구하기
max_rating = df['IMDB_Rating'].max()
print("최고 평점: ", max_rating)

최고 평점:  9.3


In [13]:
# 불리언 인덱싱(boolean indexing)을 통하여 최대평점 영화 찾기
highest_rated_movie = df[df['IMDB_Rating'] == max_rating]
print("최고 평점 영화의 제목과 평균:\n", highest_rated_movie[['Series_Title', 'IMDB_Rating']])

최고 평점 영화의 제목과 평균:
                Series_Title  IMDB_Rating
0  The Shawshank Redemption          9.3


### Chapter 7-3-3 최저 평점 영화 찾기 <a class="anchor" id="chapter7-3-3"></a>

In [14]:
# IMDB 영화 중 최저 평점 구하기
min_rating = df['IMDB_Rating'].min()
print("최저 평점: ", min_rating)

최저 평점:  7.6


In [15]:
# 불리언 인덱싱(boolean indexing)을 통하여 최저평점 영화 찾기
lowest_rated_movie = df[df["IMDB_Rating"] == min_rating]
print("최저 평점 영화의 제목과 평균:\n", lowest_rated_movie[['Series_Title', 'IMDB_Rating']])

최저 평점 영화의 제목과 평균:
                          Series_Title  IMDB_Rating
877                       Dark Waters          7.6
878                         Searching          7.6
879  Once Upon a Time... in Hollywood          7.6
880                          Nelyubov          7.6
881               The Florida Project          7.6
..                                ...          ...
995            Breakfast at Tiffany's          7.6
996                             Giant          7.6
997             From Here to Eternity          7.6
998                          Lifeboat          7.6
999                      The 39 Steps          7.6

[123 rows x 2 columns]


In [16]:
# 최저 평점을 받은 영화의 수
print("최저 평점을 받은 영화의 수: ", len(lowest_rated_movie))

최저 평점을 받은 영화의 수:  123


### Chapter 7-3-4 전체 평가 참여자 수와 영화당 평균 참여자 수 구하기 <a class="anchor" id="chapter7-3-4"></a>

In [3]:
# No_of_Votes 열의 합을 구한다.
total_votes = df['No_of_Votes'].sum()
print("총 투표 수: ", total_votes)

총 투표 수:  273692911


In [4]:
# No_of_Votes 열의 평균 계산
average_votes = df['No_of_Votes'].mean()
print("평균 투표 수: ", average_votes)

평균 투표 수:  273692.911


### Chapter 7-3-5 데이터 분석 결과 <a class="anchor" id="chapter7-3-5"></a>
1. IMDB 데이터세의 평점과 참여주 수 분석 결과는 다음과 같다.
    - 평균 평점: 7.95점
    - 최고 평점 영화: The Shawshank Redemption (9.3점)
    - 최저 평점 영화: 7.6점
    - 전체 평가 참여자 수: 1,000,000명
    - 영화당 평균 참여자 수: 273,639,911명

## Chapter 7-4 셀프체크 <a class="anchor" id="chapter7-4"></a>

In [6]:
# IMDB 데이터 중 참여자 수가 평균 이상인 영화면 필터링해 총 몇편인지 계산
# 참여자 평균 구하기
average_vote = df['No_of_Votes'].mean()
print("average_vote : ",average_vote)

over_average_vote_movis = df[df['No_of_Votes'] >= average_vote];
print(len(over_average_vote_movis))

average_vote :  273692.911
327


In [23]:
# IMDB 전체 영화의 평균 수익
# 결측치 계산하기
missing_values = df['Gross'].isnull().sum()

# 결측치가 있는 행 제거
df_without_missing = df.dropna(subset=['Gross'])

# "," 제거하하기
df_without_missing['Gross'] = df_without_missing['Gross'].str.replace(',','')

# Gross 열을 숫자형으로 변환하기
df_without_missing['Gross'] = df_without_missing['Gross'].astype(float)

# Gross 열의 평균 계산하기
average_gross = df_without_missing['Gross'].mean()
print("IMDB 전체 영화의 평균 수익: ", average_gross)


IMDB 전체 영화의 평균 수익:  68034750.87364621


In [24]:
# IMDB 전체 영화의 평균 수익
# "," 제거하하기
df['Gross'] = df['Gross'].str.replace(',','')

# 숫자로 변환 불가능한 값 NnN처리
df['Gross'] = pd.to_numeric(df['Gross'], errors='coerce')

# 결측치를 제거하고 평균 구하기
average_gross = df['Gross'].dropna().mean()
print("IMDB 전체 영화의 평균 수익: ", average_gross)

IMDB 전체 영화의 평균 수익:  68034750.87364621


In [28]:
# IMDB 영화 중 수익이 가장 높은 영화의 제목과 수익 출력하기
# 가장 높은 수익 구하기
max_gross = df['Gross'].max()

# 수익이 가장높은 영화의 제목과 수익 출력
df_max_gross = df[df['Gross'] == max_gross ]
df_max_gross[['Series_Title', 'Gross']]


,Series_Title,Gross
477,Star Wars: Episode VII - The Force Awakens,936662225.0
